util: Add [[nodiscard]] to all {Decode,Parse}[...](...) functions returning bool

practicalswift opened the pull request:
Changes in this PR:
* ~~Add linter to make sure the return value of `Parse[...](...)` is checked~~
* Add `__attribute__((warn_unused_result))` to all `{Decode,Parse}[...](...)` functions returning `bool`
* Fix violations

Context:
* #13712: `wallet: Fix non-determinism in ParseHDKeypath(...). Avoid using an uninitialized variable in path calculation.` would have been prevented by this

practicalswift committed:
miner: Default to DEFAULT_BLOCK_MIN_TX_FEE if unable to parse -blockmintxfee

fanquake commented:
```
$ test/lint/lint-all.sh
Missing "export LC_ALL=C" (to avoid locale dependence) as first non-comment non-empty line in test/lint/lint-cpp.sh
^---- failure generated from test/lint/lint-shell-locale.sh
The command "test/lint/lint-all.sh" exited with 1.
```

Empact commented on src/miner.cpp:
```diff
@@ -72,8 +72,11 @@ static BlockAssembler::Options DefaultOptions()
+            options.blockMinFeeRate = CFeeRate(n);
+        } else {
+            options.blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
+        }
```
You could avoid the duplicate else clause by boosting amount out like so:

```C++
if (gArgs.IsArgSet("-blockmintxfee") && ParseMoney(gArgs.GetArg("-blockmintxfee", ""), n)) {
```

practicalswift commented:
@fanquake Thanks! `lint-shell-locale.sh` to the rescue once again! Now fixed! Please re-review :-)

@Empact Good point! Updated. Please re-review :-)

skeees commented:
utACK (weakly regarding the linter)
If you want to have lots of lint checks and similar static analysis to the kind you are adding here (it seems you do, and these things are useful!) - it might make sense to consider investing in building out tooling that uses libclang and inspects the AST directly as opposed to just regexp matching.

practicalswift commented:
@skeees I removed the linter. Now using `__attribute__((warn_unused_result))` instead. Please re-review :-)

skeees commented:
:) Nice
Is is best practice to `#define` this attribute similar to how this is done in `threadsafety.h` or `#  define SECP256K1_WARN_UNUSED_RESULT __attribute__ ((__warn_unused_result__))` ?

practicalswift committed:
tests: Check return value of ParseParameters(...)

practicalswift commented:
@skeees The `secp256k1` define is used to handle ancient `gcc` versions AFAIC, so I see no reason to use that here. Let me know if you see any other reason :-)

practicalswift committed:
tests: Explicitly ignore the return value of DecodeBase58(...)

practicalswift commented:
Also added `__attribute__((warn_unused_result))` to all Decode[...](...) functions returning `bool`. Please re-review :-)

skeees commented:
can reorder the Tests* commits first before adding the attribute commits so that you don't break a bisect

practicalswift commented:
@skeees When rethinking it I think you have a good point regarding using a `define`. Perhaps we should go with:

```
#ifdef __has_cpp_attribute
#  if __has_cpp_attribute(nodiscard)
#    define NODISCARD [[nodiscard]]
#  endif
#else
#    define NODISCARD __attribute__((warn_unused_result))
#endif
```

That will have the benefit of being less verbose. That would use the newer `[[nodiscard]]` (required from C++17) where available and fall back on the older `__attribute__((warn_unused_result))`.

practicalswift commented:
@skeees Commits reordered as requested!

practicalswift commented:
What about the following?

```
$ cat src/nodiscard.h
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2018 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#ifndef BITCOIN_NODISCARD_H
#define BITCOIN_NODISCARD_H

#ifdef __has_cpp_attribute
#  if __has_cpp_attribute(nodiscard)
#    define NODISCARD [[nodiscard]]
#  endif
#else
#  define NODISCARD __attribute__((warn_unused_result))
#endif

#endif // BITCOIN_NODISCARD_H
```

Empact commented:
@practicalswift You need to add `attributes.h` to `src/Makefile.am`.

Empact commented on src/attributes.h:
```diff
@@ -0,0 +1,17 @@
+#define BITCOIN_ATTRIBUTES_H
+
+#ifdef __has_cpp_attribute
+#  if __has_cpp_attribute(nodiscard)
```
nit: could collapse this to one level: `#if defined(__has_cpp_attribute) && __has_cpp_attribute(nodiscard)`

practicalswift commented:
@Empact Thanks for reviewing! Feedback addressed. The broken `ifdef` logic was sloppy from my side – sorry about that. Would you mind re-reviewing? :-)

Empact commented:
On second thought, how about throw rather than default on unparsable option? Seems like a failure to be noisy about.

skeees commented:
In terms of commit ordering, you need the `#define NODISCARD` commit to precede the commits where you use `NODISCARD` Other than that lgtm!

practicalswift commented:
Rebased! :-)

@MarcoFalke I don't know if I've thanked you before but thanks for your work on @DrahtBot! It really helps keeping the PR:s up to date! Really nice!

practicalswift commented:
Updated `NODISCARD` to work also with Visual C++ :-)

practicalswift commented:
Rebased!

arvidn commented on src/rest.cpp:
```diff
@@ -105,7 +106,7 @@ static std::string AvailableDataFormatsString()
 }
 
-static bool ParseHashStr(const std::string& strReq, uint256& v)
+NODISCARD static bool ParseHashStr(const std::string& strReq, uint256& v)
```
ditto for all other definitions

Empact replied:
This is misleading because it is actually a static function, there is no declaration, though there is another identically-named function. See #14307

practicalswift commented on src/miner.cpp:
```diff
@@ -70,9 +70,8 @@ static BlockAssembler::Options DefaultOptions()
-        CAmount n = 0;
-        ParseMoney(gArgs.GetArg("-blockmintxfee", ""), n);
+    CAmount n = 0;
+    if (gArgs.IsArgSet("-blockmintxfee") && ParseMoney(gArgs.GetArg("-blockmintxfee", ""), n)) {
```
@arvidn We allow only C++11 :-)

Empact commented on src/utilstrencodings.h:
```diff
@@ -35,8 +37,8 @@ enum SafeChars
-std::vector<unsigned char> ParseHex(const char* psz);
-std::vector<unsigned char> ParseHex(const std::string& str);
+NODISCARD std::vector<unsigned char> ParseHex(const char* psz);
+NODISCARD std::vector<unsigned char> ParseHex(const std::string& str);
```
These don't return `bool`

Empact commented:
I see apparently unnecessary changes to:
src/wallet/rpcwallet.cpp
src/wallet/test/psbt_wallet_tests.cpp

practicalswift committed:
Add NODISCARD to all {Decode,Parse}[...](...) functions returning bool. Sort includes.

practicalswift commented:
@Empact Unrelated sorting of includes in those two files removed :-)

ken2812221 commented on src/Makefile.am:
```diff
@@ -92,6 +92,7 @@ endif
 BITCOIN_CORE_H = \
   addrdb.h \
   addrman.h \
+  attributes.h \
```
Wrong file, should be `nodiscard.h`?

practicalswift replied:
`attributes.h` is correct but that file got lost in the last rebase. Now re-added :-)

ken2812221 commented on src/base58.h:
```diff
@@ -14,6 +14,8 @@
 #ifndef BITCOIN_BASE58_H
 #define BITCOIN_BASE58_H
 
+#include <attributes.h>
```
Same here

practicalswift replied:
`attributes.h` is correct but that file got lost in the last rebase. Now re-added :-)

ken2812221 replied:
Seems `attributes.h` is as same as `nodiscard.h` now

practicalswift replied:
`nodiscard.h` now removed :-)

practicalswift commented on src/script/descriptor.cpp:
```diff
@@ -431,7 +431,7 @@ std::vector<Span<const char>> Split(const Span<const char>& sp, char sep)
 
 /** Parse a key path, being passed a split list of elements (the first element is ignored). */
-bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out)
+NODISCARD static bool ParseKeyPath(const std::vector<Span<const char>>& split, KeyPath& out)
```
Yes, using `static` here for consistency and keeping the number of lines touched to a minimum.

practicalswift commented:
Rebased! Please re-review @ken2812221, @Empact, @skeees and @fanquake :-)

practicalswift commented:
Updated to address @arvidn:s `static` nit. Please re-review :-)

ken2812221 commented on src/attributes.h:
```diff
@@ -0,0 +1,17 @@
+#else
+#  define NODISCARD __attribute__((warn_unused_result))
+#endif
+
```
This solution might be too complex, but it can get rid of annoying MSVC warnings.
```cpp
#if defined(__has_cpp_attribute)
#  if __has_cpp_attribute(nodiscard)
#    define NODISCARD [[nodiscard]]
#  endif
#endif
#ifndef NODISCARD
#  if defined(_MSC_VER) && _MSC_VER >= 1700
#    define NODISCARD _Check_return_
#  else
#    define NODISCARD __attribute__((warn_unused_result))
#  endif
#endif
```

practicalswift replied:
Fixed!

practicalswift commented:
Updated to address MSVC warnings. Please re-review :-)

practicalswift commented:
Rebased!

MarcoFalke commented:
utACK 9cc0230cfc1ae9b9c1c905cd9ac613bc98bfa43a

Empact commented:
utACK 9cc0230

sipa commented:
utACK 9cc0230cfc1ae9b9c1c905cd9ac613bc98bfa43a
